src/backtrace/dbghelp.rs

```diff
@@ -34,62 +36,108 @@ unsafe impl Sync for Frame {}
+        self.inline_context
     }
+}
 
-    fn addr_pc_mut(&mut self) -> &mut ADDRESS64 {
-        match self.stack_frame {
-            StackFrame::New(ref mut new) => &mut new.AddrPC,
-            StackFrame::Old(ref mut old) => &mut old.AddrPC,
-        }
+#[repr(C, align(16))] // required by `CONTEXT`, is a FIXME in winapi right now
+struct MyContext(CONTEXT);
+
+#[cfg(target_arch = "x86_64")]
+impl MyContext {
+    #[inline(always)]
+    fn ip(&self) -> DWORD64 {
+        self.0.Rip
     }
 
-    fn addr_frame_mut(&mut self) -> &mut ADDRESS64 {
-        match self.stack_frame {
-            StackFrame::New(ref mut new) => &mut new.AddrFrame,
-            StackFrame::Old(ref mut old) => &mut old.AddrFrame,
-        }
+    #[inline(always)]
+    fn sp(&self) -> DWORD64 {
+        self.0.Rsp
     }
+}
 
-    fn addr_stack(&self) -> &ADDRESS64 {
-        match self.stack_frame {
-            StackFrame::New(ref new) => &new.AddrStack,
-            StackFrame::Old(ref old) => &old.AddrStack,
-        }
+#[cfg(target_arch = "aarch64")]
+impl MyContext {
+    #[inline(always)]
+    fn ip(&self) -> DWORD64 {
+        self.0.Pc
     }
 
-    fn addr_stack_mut(&mut self) -> &mut ADDRESS64 {
-        match self.stack_frame {
-            StackFrame::New(ref mut new) => &mut new.AddrStack,
-            StackFrame::Old(ref mut old) => &mut old.AddrStack,
-        }
+    #[inline(always)]
+    fn sp(&self) -> DWORD64 {
+        self.0.Sp
     }
 }
 
-#[repr(C, align(16))] // required by `CONTEXT`, is a FIXME in winapi right now
-struct MyContext(CONTEXT);
+#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
+#[inline(always)]
+pub unsafe fn trace(cb: &mut dyn FnMut(&super::Frame) -> bool) {
+    use core::ptr;
+
+    let mut context = core::mem::zeroed::<MyContext>();
+    RtlCaptureContext(&mut context.0);
+
+    // Call `RtlVirtualUnwind` to find the previous stack frame, walking until we hit ip = 0.
+    while context.ip() != 0 {
+        let mut base = 0;
+
+        let fn_entry = RtlLookupFunctionEntry(context.ip(), &mut base, ptr::null_mut());
+        if fn_entry.is_null() {
+            break;
+        }
 
+        let frame = super::Frame {
+            inner: Frame {
+                base_address: fn_entry as _,
+                ip: context.ip() as *mut _,
+                sp: context.sp() as *mut _,
+                #[cfg(not(target_env = "gnu"))]
+                inline_context: None,
+            },
+        };
+
+        if !cb(&frame) {
+            break;
+        }
+
+        let mut handler_data = 0usize;
+        let mut establisher_frame = 0;
+
+        RtlVirtualUnwind(
+            0,
+            base,
+            context.ip(),
+            fn_entry,
+            &mut context.0,
+            &mut handler_data as *mut _ as *mut _,
+            &mut establisher_frame,
+            ptr::null_mut(),
+        );
+    }
+}
+
+#[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
```

Comment:
Yeah, the behavior I wanted was "for `x86_64` and `aarch64`, use the function on line 91, for all others use this one" so that if another arch was ever supported, we'd at least fall back to this implementation. 

If you'd prefer I spell out `x86` and `arm` instead (I believe those are the only relevant archs in question), I can certainly do so! 🙂

Reply:
Tbh, what particularly brought it to mind was the line further down that explicitly uses `IMAGE_FILE_MACHINE_I386` (which is fine for arm too but made me pause for a sec). However, ultimately I do see the value in having a fallback so I'm fine with leaving the `cfg` as is.

Reply:
Oh, that's fair! I forgot about that bit. Any new archs that get added to Windows will almost certainly support `RtlVirtualUnwind` so the set of archs that will need to fallback to using `StackWalkEx`/`StackWalk64` should basically be complete at this time. 

I'll push a commit that updates the `cfg` to make that explicit and also fix the `IMAGE_FILE_MACHINE_I386` for arm32.

Reply:
Done

Reply:
Great, thanks!